ykcompile/src/lib.rs

```diff
@@ -250,41 +258,99 @@ impl TraceCompiler {
             ));
         };
 
-        // FIXME implement support for arguments.
-        if !args.is_empty() {
-            return Err(CompileError::Unimplemented("call with args".to_owned()));
+        if args.len() > 6 {
+            return Err(CompileError::Unimplemented(
+                "call with spilled args".to_owned(),
+            ));
         }
 
-        // Generally speaking, we don't know what the callee will clobber, so we save our registers
-        // before doing the call. For now we assume that the caller is OK with RAX being clobbered
-        // though.
-        for reg in 1..X86_86_NUM_REGS {
+        // Figure out where the return value (if there is one) is going.
+        let dest_reg: Option<u8> = if let Some(d) = dest {
+            Some(self.local_to_reg(d.local)?)
+        } else {
+            None
+        };
+
+        // Save Sys-V callee save registers to the stack, but skip the one (if there is one) that
+        // will store the return value. It's safe to assume the caller expects this to be
+        // clobbered.
+        //
+        // FIXME: Note that we don't save rax. Although this is a callee save register, the way the
+        // tests currently work is they check the last value returned at the end of the trace. This
+        // value is assumed to remain in rax. If we were to restore rax, we'd break that. Note that
+        // the register allocator never gives out rax for this precise reason.
+        //
+        // rdi, rsi, rdx, rcx, r8, r9, r10, r11.
+        let save_regs = [7, 6, 2, 1, 8, 9, 10, 11]
+            .iter()
+            .filter(|r| Some(**r) != dest_reg)
+            .map(|r| *r)
+            .collect::<Vec<u8>>();
+        for reg in &save_regs {
             dynasm!(self.asm
                 ; push Rq(reg)
             );
         }
 
+        let stack_index = |reg: u8| -> i32 {
+            i32::try_from(save_regs.iter().rev().position(|&r| r == reg).unwrap()).unwrap()
```

Comment:
I think this could benefit from a comment as it's meaning is difficult to unravel. For example, why is it beneficial here to reverse the iterator? Do we know that `reg` can always be found towards the end of the vec?

Reply:
It's due to the way the registers will be laid out on a stack. The first element in `save_regs` will be at the highest stack offset, because it will be pushed first.

Will add a comment.